staging/src/k8s.io/apimachinery/pkg/labels/selector_test.go

```diff
@@ -303,27 +309,78 @@ func TestRequirementConstructor(t *testing.T) {
-		{"y", selection.GreaterThan, sets.NewString("1"), true},
-		{"z", selection.LessThan, sets.NewString("6"), true},
-		{"foo", selection.GreaterThan, sets.NewString("bar"), false},
-		{"barz", selection.LessThan, sets.NewString("blah"), false},
-		{strings.Repeat("a", 254), selection.Exists, nil, false}, //breaks DNS rule that len(key) <= 253
+		{"x1", selection.In, nil, &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values",
+			BadValue: []string{},
+		}},
+		{"x2", selection.NotIn, sets.NewString(), &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values",
+			BadValue: []string{},
+		}},
+		{"x3", selection.In, sets.NewString("foo"), nil},
+		{"x4", selection.NotIn, sets.NewString("foo"), nil},
+		{"x5", selection.Equals, sets.NewString("foo", "bar"), &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values",
+			BadValue: []string{"bar", "foo"},
+		}},
+		{"x6", selection.Exists, nil, nil},
+		{"x7", selection.DoesNotExist, nil, nil},
+		{"x8", selection.Exists, sets.NewString("foo"), &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values",
+			BadValue: []string{"foo"},
+		}},
+		{"x9", selection.In, sets.NewString("bar"), nil},
+		{"x10", selection.In, sets.NewString("bar"), nil},
+		{"x11", selection.GreaterThan, sets.NewString("1"), nil},
+		{"x12", selection.LessThan, sets.NewString("6"), nil},
+		{"x13", selection.GreaterThan, nil, &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values",
+			BadValue: []string{},
+		}},
+		{"x14", selection.GreaterThan, sets.NewString("bar"), &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values[0]",
+			BadValue: "bar",
+		}},
+		{"x15", selection.LessThan, sets.NewString("bar"), &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values[0]",
+			BadValue: "bar",
+		}},
+		{strings.Repeat("a", 254), selection.Exists, nil, &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "key",
+			BadValue: strings.Repeat("a", 254),
+		}}, //breaks DNS rule that len(key) <= 253
+		{"x16", selection.Equals, sets.NewString(strings.Repeat("a", 254)), &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values[0]",
+			BadValue: strings.Repeat("a", 254),
+		}},
+		{"x17", selection.Equals, sets.NewString("a b"), &field.Error{
+			Type:     field.ErrorTypeInvalid,
+			Field:    "values[0]",
+			BadValue: "a b",
+		}},
+		{"x18", "unsupportedOp", nil, &field.Error{
+			Type:     field.ErrorTypeNotSupported,
+			Field:    "operator",
+			BadValue: selection.Operator("unsupportedOp"),
+		}},
 	}
 	for _, rc := range requirementConstructorTests {
-		if _, err := NewRequirement(rc.Key, rc.Op, rc.Vals.List()); err == nil && !rc.Success {
-			t.Errorf("expected error with key:%#v op:%v vals:%v, got no error", rc.Key, rc.Op, rc.Vals)
-		} else if err != nil && rc.Success {
+		_, err := NewRequirement(rc.Key, rc.Op, rc.Vals.List())
+		if err != nil && rc.WantErr == nil {
```

Comment:
This should already be handled by diff

Reply:
I add this because I just found diff handling nil not very well, it will output a strange log when getting a nil value, for example (change x18 wantErr to nil):
```
=== RUN   TestRequirementConstructor
    selector_test.go:458: NewRequirement test x18 returned unexpected error (-want,+got):
          interface{}(
        + 	e`operator: Unsupported value: "unsupportedOp": supported values: "in", "notin", "=", "==", "!=", "exists", "!", "gt", "lt"`,
          )
```
It doesn't indicate the wanted value is nil and lost type information too.